### ISSUE TITLE
How can I load customized evaluation metric for llama with SFTtrainer?

### ISSUE THREAD

**Issue by daehuikim** [huggingface/trl#893]
Hello , Sorry for noob question. I am dealing with llama2 fine tuning using trl & peft.
I just found default evaluation metric is based on recall.
So, I tried to change evaluation metric However, It does not work for even in A100-80GB.
Whenever trainer try evaluating, OOM error occurs with my customized bleu metric.
If it's possible, I want to know how to adapt my idea to SFTtrainer with some exapmle codes please.

Below are my code.
(WIthout compute_metric_bleu, everything is ok)

```
def compute_metrics_bleu(pred):
    references = pred.label_ids
    generated_texts = pred.predictions
    print("references:",references)
    print("generated_texts:",generated_text)
    
    bleu_scores = []
    for reference, generated_text in zip(references, generated_texts):
        reference_text = train_dataset[reference]['text']
        bleu_score = sentence_bleu([reference_text], generated_text)
        bleu_scores.append(bleu_score)

    return {
        'bleu': sum(bleu_scores) / len(bleu_scores)
    }

(SKIP training argument & Lora Config)


    trainer = SFTTrainer(
        model=model,
        train_dataset=train_dataset,
        eval_dataset=valid_dataset,  # Pass validation dataset here
        peft_config=peft_config,
        dataset_text_field="text",
        max_seq_length=max_seq_length,
        tokenizer=tokenizer,
        args=training_arguments,
        packing=packing,
        compute_metrics = compute_metrics_bleu
    )

    trainer.train()
    trainer.model.save_pretrained(new_model)
```

**Comment by SingL3** [comment]
Hi, the eval process in the `Trainer` of `transformers` would gather all the `logits` whose size is `batch_size x seq_len x vocab_size`, and for most of the time the `vocab_size` is like 32k or more, thus cosuming lots of memory.
There is `preprocess_logits_for_metrics` in the args of `Trainer` to handle this situation.

**Comment by daehuikim** [comment]
> Hi, the eval process in the `Trainer` of `transformers` would gather all the `logits` whose size is `len(eval_dataset) x seq_len x vocab_size` or `batch_size x seq_len x vocab_size`, and for most of the time the `vocab_size` is like 32k or more, thus cosuming lots of memory. There is `preprocess_logits_for_metrics` in the args of `Trainer` to handle this situation. Check it [here](https://huggingface.co/docs/transformers/v4.34.1/en/main_classes/trainer#transformers.Trainer.preprocess_logits_for_metrics)

Thanks for kind answer. It seems ```batch_size x seq_len x vocab_size``` doesnt changed because I just changed my ```compute_metrics``` argument. Can it be matter even in this case? If it is how can i debug or edit the code properly. sorry for noob question again.

**Comment by SingL3** [comment]
@daehuikim 
In my case, I need to compute the tokens acc of the generation, so I define a `preprocess_logits_for_metrics ` function to get the `argmax` of the `logits`. And I would get an result tensor with size `batch_size x seq_len`, which is much smaller.
In this case, I would have to pass this function as `preprocess_logits_for_metrics`:
```
from typing import Optional
import torch

def logits_argmax(logits: torch.Tensor, labels: Optional[torch.Tensor] = None):
    if isinstance(logits, tuple):
        # Depending on the model and config, logits may contain extra tensors,
        # like past_key_values, but logits always come first
        logits = logits[0]
    return logits.argmax(dim=-1)
```

**Comment by daehuikim** [comment]
@SingL3 
You gave me excellent example for me. I am doing for acc of the generation either. I am still curious why ```logits = logits[0]``` does not damage to my logits.
